staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go

```diff
@@ -710,6 +713,82 @@ func TestApplyPruneObjects(t *testing.T) {
 	}
 }
 
+// Tests that apply of object in need of CSA migration results in a call
+// to patch it.
+func TestApplyCSAMigration(t *testing.T) {
```

Comment:
This is a very basic unit test. I am unsure how kubectl testing is organized completely. 

I'd like to add a test that shows a client-side-apply followed by a server-side-apply does not result in undeleteable fields. Need an e2e test for that I assume

Reply:
integration tests would be very useful for this fix.

Reply:
There's also a series of tests available in https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/test-cmd.sh, which is a little messy and hard to debug, but can be used to test what you want I think.

Reply:
We'll need to add significant unit tests. Integration tests (including `make test-cmd`) are helpful, but the focus must be on the unit tests.